core/src/main/java/tc/oc/pgm/loot/LootableMatchModule.java

```diff
@@ -0,0 +1,337 @@
+              .onRise(
+                  Match.class,
+                  filler.getRefillTrigger(),
+                  m -> this.triggeredFillerInvalidator.accept(filler));
+        });
+  }
+
+  /**
+   * Return a predicate that applies a Filter to the given InventoryHolder, or null if the
+   * InventoryHolder is not something that we should be filling.
+   */
+  private static @Nullable Predicate<Filter> passesFilter(InventoryHolder holder) {
+    if (holder instanceof DoubleChest) {
+      final DoubleChest doubleChest = (DoubleChest) holder;
+      return filter ->
+          !filter.query(new BlockQuery((Chest) doubleChest.getLeftSide())).isDenied()
+              || !filter.query(new BlockQuery((Chest) doubleChest.getRightSide())).isDenied();
+    } else if (holder instanceof BlockState) {
+      return filter -> !filter.query(new BlockQuery((BlockState) holder)).isDenied();
+    } else if (holder instanceof Player) {
+      // This happens with crafting inventories, and possibly other transient inventory types
+      // Pretty sure we never want to fill an inventory held by the player
+      return null;
+    } else if (holder instanceof Entity) {
+      return filter -> !filter.query(new EntityQuery((Entity) holder)).isDenied();
+    } else {
+      // If we're not sure what it is, don't fill it
+      return null;
+    }
+  }
+
+  @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
+  public void onInventoryOpen(InventoryOpenEvent event) {
+    final MatchPlayer opener = this.match.getParticipant(event.getActor());
+    if (opener == null) return;
+
+    final Inventory inventory = event.getInventory();
+    final Predicate<Filter> passesFilter = passesFilter(inventory.getHolder());
+    if (passesFilter == null) return;
+
+    logger.fine(
+        () -> opener.getName() + " opened a " + inventory.getHolder().getClass().getSimpleName());
+
+    // Find all Fillers that apply to the holder of the opened inventory
+    final List<FillerDefinition> fillers =
+        this.fillers.stream()
+            .filter(filler -> passesFilter.test(filler.getFilter()))
+            .collect(Collectors.toList());
+    if (fillers.isEmpty()) return;
+
+    logger.fine(
+        () ->
+            "Found fillers "
+                + fillers.stream()
+                    .map(FillerDefinition::toString)
+                    .collect(Collectors.joining(", ")));
+
+    // Find all Caches that the opened inventory is part of
+    final List<Fillable> fillables = new ArrayList<>();
+    for (Cache cache : caches) {
+      if (passesFilter.test(cache.region()) && passesFilter.test(cache.filter())) {
+        fillables.add(new FillableCache(cache));
+      }
+    }
+    // If the inventory is not in any Cache, just fill it directly
+    if (fillables.isEmpty()) {
+      fillables.add(new FillableInventory(inventory));
+    }
+
+    fillables.forEach(fillable -> fillable.fill(opener, fillers));
```

Comment:
```suggestion
    // Find all Caches that the opened inventory is part of
    boolean foundCache = false;
    for (Cache cache : caches) {
      if (passesFilter.test(cache.region()) && passesFilter.test(cache.filter())) {
        foundCache = true;
        new FillableCache(cache).fill(opener, fillers);
      }
    }
    // If the inventory is not in any Cache, just fill it directly
    if (!foundCache) {
      new FillableInventory(inventory).fill(opener, fillers);
    }
```
Any reason why you'd need this the other way around? It does introduce a behavior change, mainly, that the tests for the filter are no longer done all before starting to fill, but it prevents needing the list.

If the order of operations is relevant, then keep it as is

Reply:
If the item filters(e.g. CarryingItemFilter) every get expanded to include entities I can see this change mattering, but it would not change legacy behavior as far as I can see. Think it's worth keeping this behavior for that potential change?

Reply:
Bump, @Pablete1234 ?